api/envoy/service/ext_proc/v3alpha/external_processor.proto

```diff
@@ -0,0 +1,300 @@
+    // and end further filter processing on the request path.
+    ImmediateResponse immediate_response = 7;
+  }
+
+  // Optional metadata that will be emitted as dynamic metadata to be consumed by the next
+  // filter. This metadata will be placed in the namespace "envoy.filters.http.ext_proc".
+  google.protobuf.Struct dynamic_metadata = 8;
+
+  // Override how parts of the HTTP request and response are processed
+  // for the duration of this particular request/response only. Servers
+  // may use this to intelligently control how requests are processed
+  // based on the headers and other metadata that they see.
+  envoy.extensions.filters.http.ext_proc.v3alpha.ProcessingMode mode_override = 9;
+}
+
+// The following are messages that are sent to the server.
+
+// This message is sent to the external server when the HTTP request and responses
+// are first received.
+message HttpHeaders {
+  // The HTTP request headers. All header keys will be
+  // lower-cased, because HTTP header keys are case-insensitive.
+  config.core.v3.HeaderMap headers = 1;
+
+  // The values of properties selected by the "request_properties"
+  // list in the configuration. Each entry in the map is the
+  // value of the specified property.
+  map<string, google.protobuf.Struct> properties = 2;
+
+  // If true, then there is no message body associated with this
+  // request or response.
+  bool end_of_stream = 3;
+}
+
+// This message contains the message body that Envoy sends to the external server.
+message HttpBody {
+  bytes body = 1;
+
+  bool end_of_stream = 2;
+}
+
+// This message contains the trailers.
+message HttpTrailers {
+  config.core.v3.HeaderMap trailers = 1;
+}
+
+// The following are messages that may be sent back by the server.
+
+// This message must be sent in response to a request_headers or response_headers
+// message.
+// [#next-free-field: 6]
+message HeadersResponse {
+  // Instructions on how to manipulate the headers
+  HeaderMutation header_mutation = 1;
+
+  // If set, add a body to the message if none was present
+  // before.
+  BodyMutation body_mutation = 2;
+
+  // If set, add trailers to the message. If and only if this
+  // field is combined with the end_stream flag, then
+  // the trailers will be sent along with the request or
+  // response -- otherwise, this field is ignored.
+  config.core.v3.HeaderMap trailers = 3;
+
+  // If true, then replace the request and response with the contents
+  // of this message. If header_mutation is set, apply it to the
+  // headers. If body_mutation is set and contains a body, then add that
+  // body to the request or response, even if one does not already exist --
+  // otherwise, clear the body. Any additional body and trailers
+  // received from downstream will be ignored, and any additional messages
+  // from the remote server on either the request or response path will
+  // also be ignored for the duration of this request or response.
+  bool end_of_stream = 4;
```

Comment:
I think you probably want this to be a WKT so you have a tri-state. The reason for that is that I think you also want to be able to set end_stream to false, and make a headers only request/response into a request/response with data, by sending data/trailers later. I commented on this in the other PR. I don't think we have to support this advanced functionality on day 1, but it would be nice to not have to deprecate things to do it. The way I think it would end up working is end_of_stream would be explicitly set to false (as opposed to empty being no change), and then the server could stream a sequence of body messages and an optional trailer message to finish the request/response. Let me know if this makes sense.

Reply:
I don't mind adding the tri-state, or perhaps another flag to make the protocol more general -- let me think about it.

In the meantime, here's how this particular use case might work:

1. onRequestHeaders is called and the "end_stream" flag is set.
2. filter sends the request_headers message on the gRPC stream.
3. Filter returns with StopAllIterationAndWatermark. (Because we don't know what the remote server wants yet we can't return a different code IMO.)
4. Filter receives a message on the stream with a status of "end_stream_later" or whatever we decide to call it
5. Filter continues to wait for additional responses with an "end_stream_finally" message. In the meantime, if it receives multiple message body messages, it buffers them.
6. Filter gets the "end_stream_finally" message and then calls continueDecoding() to resume the filter chain

Note that this is different from the ContinueAndDontEndStream status that's in the HTTP filter interface, because with that interface, we continue processing the filter chain and wait for the new data, whereas here we can't continue processing in that way because we already returned a different status and the filter interface gives us no way to control what "continueDecoding()" means.

I might argue that unless the filter processing options change, we could instead roll up steps 3-5 there into a single message -- let the remote server do all that work and respond only when it's ready for Envoy to continue processing.

Reply:
I think you have the right idea, but I think it's possible to also stream data from the ext proc server without buffering, and I could see this being desired in advanced use cases. See the "inject" family of functions in filter.h.